internal/servicelog/iterator.go

```diff
@@ -0,0 +1,167 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>
+
+package servicelog
+
+import (
+	"io"
+)
+
+type Iterator interface {
+	// Close closes the iterator so that buffers can be used for future writes.
+	// If Close is not called, the iterator will block buffer recycling causing
+	// write failures.
+	Close() error
+	// Next advances to the next available buffered write.
+	// When passed cancel, Next will wait for the next buffered write.
+	// If a nil channel is passed, cancel will return the first available
+	// buffered write, if none is available, return immediatley.
+	Next(cancel <-chan struct{}) bool
+
+	Buffered() int
+	io.Reader
+	io.WriterTo
+}
+
+type iterator struct {
+	rb           *RingBuffer
+	index        RingPos
+	trunc        []byte
+	truncWritten bool
+	notifyChan   chan struct{}
+	closeChan    chan struct{}
+}
+
+var _ Iterator = (*iterator)(nil)
+
+var (
+	truncBytes = []byte("<trunc>\n")
+)
+
+func (it *iterator) Close() error {
+	if it.rb == nil {
+		return nil
+	}
+	it.rb.removeIterator(it)
+	close(it.notifyChan)
+	it.rb = nil
+	return nil
+}
+
+func (it *iterator) Next(cancel <-chan struct{}) bool {
+	if it.rb == nil {
+		return false
+	}
+	select {
+	case <-it.notifyChan:
+	default:
+	}
+	start, end := it.rb.Positions()
+	if it.index < start {
+		it.index = start
+		it.truncated()
+	}
+	if it.index < end || len(it.trunc) > 0 {
+		return true
+	}
+	for cancel != nil {
+		// if passed a cancel channel, wait for more data.
+		closed := false
+		select {
+		case <-it.closeChan:
+			closed = it.rb.Closed()
+		case <-cancel:
+			cancel = nil
+		case <-it.notifyChan:
+		}
+		start, end := it.rb.Positions()
+		if it.index < start {
+			it.index = start
+			it.truncated()
+		}
+		if it.index < end || len(it.trunc) > 0 {
+			return true
+		}
+		if it.index == end && closed {
+			cancel = nil
+		}
+	}
+	return false
+}
+
+// Read implements io.Reader
+func (it *iterator) Read(dest []byte) (int, error) {
+	if it.rb == nil {
+		return 0, io.EOF
+	}
+	if len(it.trunc) > 0 {
+		n := copy(dest, it.trunc)
+		it.trunc = it.trunc[n:]
+		it.truncWritten = true
+		return n, nil
+	}
+	n, err := it.rb.Copy(dest, it.index)
+	if err == ErrRange {
+		it.truncated()
+		err = nil
+	}
+	if n > 0 {
+		it.truncWritten = false
+	}
+	it.index += RingPos(n)
+	return n, err
+}
+
+// WriteTo implements io.WriterTo
+func (it *iterator) WriteTo(writer io.Writer) (int64, error) {
+	if it.rb == nil {
+		return 0, io.EOF
+	}
+	if len(it.trunc) > 0 {
+		n, err := writer.Write(it.trunc)
+		it.trunc = it.trunc[n:]
+		it.truncWritten = true
+		return int64(n), err
+	}
+	n, err := it.rb.WriteTo(writer, it.index)
+	if err == ErrRange {
+		it.truncated()
+		err = nil
+	}
+	if n > 0 {
+		it.truncWritten = false
+	}
+	it.index += RingPos(n)
```

Comment:
On truncation, won't this just keep writing the truncation message forever if the buffers are off? Should probably be set to the start of the current position, but even that sounds a bit suspect because there's a chance the buffer will be off again next time we try. Ideal situation is probably to make the index float on truncation, and then only redefine it once we're effectively reading some real data, so that we'll immediately move the marker forward and make room for the ring buffer write to overwrite some data without another immediate truncation.

Reply:
Good idea, I've moved this to a floating index on truncation, so that the next read will always be the tail.